test/test_dataloader.py

```diff
@@ -3673,7 +3724,1264 @@ def test_out_of_order_iterable_ds(self):
+
+    def test_reproducible_shuffle_sequences(self):
+        """Test that shuffle produces reproducible sequences when resumed"""
+        dataset = StatefulMapDataset(16, shuffle=False)
+
+        # Run full sequence with shuffling
+        generator1 = torch.Generator()
+        generator1.manual_seed(123)
+
+        dl1 = DataLoader(
+            dataset=dataset,
+            num_workers=0,
+            stateful=True,
+            batch_size=2,
+            shuffle=True,
+            generator=generator1,
+        )
+
+        full_sequence = []
+        for batch in dl1:
+            full_sequence.extend(batch)
+
+        # Now run with checkpointing at different points
+        generator2 = torch.Generator()
+        generator2.manual_seed(123)  # Same seed
+
+        dl2 = DataLoader(
+            dataset=dataset,
+            num_workers=0,
+            stateful=True,
+            batch_size=2,
+            shuffle=True,
+            generator=generator2,
+        )
+
+        # Consume first 3 batches (6 items)
+        checkpointed_sequence = []
+        it = iter(dl2)
+        for i in range(3):
+            batch = next(it)
+            checkpointed_sequence.extend(batch)
+
+        # Save state and get remaining items
+        state_dict = dl2.state_dict()
+        remaining_original = []
+        for batch in it:
+            remaining_original.extend(batch)
+
+        # Resume from checkpoint
+        generator3 = torch.Generator()
+        generator3.manual_seed(123)  # Same seed
+
+        dataset3 = StatefulMapDataset(16, shuffle=False)
+        dl3 = DataLoader(
+            dataset=dataset3,
+            num_workers=0,
+            stateful=True,
+            batch_size=2,
+            shuffle=True,
+            generator=generator3,
+        )
+        dl3.load_state_dict(state_dict)
+
+        remaining_resumed = []
+        for batch in dl3:
+            remaining_resumed.extend(batch)
+
+        # Verify sequences match
+        full_reconstructed = checkpointed_sequence + remaining_resumed
+        self.assertEqual(len(full_sequence), len(full_reconstructed))
+        for orig, recon in zip(full_sequence, full_reconstructed):
+            self.assertEqual(orig, recon)
+
+        # Also verify the remaining parts match
+        self.assertEqual(remaining_original, remaining_resumed)
+
+
+@unittest.skipIf(
+    TEST_WITH_TSAN,
+    "Fails with TSAN with the following error: starting new threads after multi-threaded "
+    "fork is not supported. Dying (set die_after_fork=0 to override)",
+)
+class TestStatefulDataLoaderMultiEpoch(TestCase):
+    """Test stateful DataLoader across multiple epochs and at epoch boundaries."""
+
+    @parametrize("num_workers", [0])
+    def test_multi_epoch_continuation(self, num_workers):
+        """Test that DataLoader state is preserved across multiple epochs"""
+        dataset = StatefulMapDataset(12, shuffle=False)
+
+        # For reproducibility
+        torch.manual_seed(789)
```

Comment:
It shouldn't be necessary to set the global seed.  You can pass a `torch.Generator` into `DataLoader`.